### PR TITLE
Calculate more statistical data

### PR DESCRIPTION
**Pull request by OrpazFishel**

### CODE REVIEW

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java [hunk 1]
```diff
@@ -36,7 +38,8 @@
     private final double transmissionThresholdRadius;
 
     private final HashMap<String, Double> distancesStats = new HashMap<>();
-    private final HashMap<String, Double> beaconsObservedSum = new HashMap<>();
+    private final HashMap<Beacon, observedInterval.Builder> beaconsObservedOpenInterval = new HashMap<>();
+    private final HashMap<Beacon, List<observedInterval>> beaconsObservedIntervals = new HashMap<>();
```
Comment (omermadmon) [inline src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:42]: Consider rename it to `beaconsObservedCurrentInterval` or add a comment instead.

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java [hunk 2]
```diff
@@ -404,7 +415,63 @@ protected AbstractSimulation (Builder builder) {
...
+        ImmutableTable.Builder<String, String, Double> observedStats = new ImmutableTable.Builder<>();
+        mapIdsToPercentOfSimulation(observedIntervals).forEach((k, v) -> observedStats.put(k, Schema.StatisticsState.observedPercent, v));
+        mapIdsToMinValue(observedIntervals).forEach((k, v) -> observedStats.put(k, Schema.StatisticsState.minimumLengthObservedInterval, v));
+        mapIdsToMinValue(unobservedIntervals).forEach((k, v) -> observedStats.put(k, Schema.StatisticsState.minimumLengthUnobservedInterval, v));
+        mapIdsToMaxValue(observedIntervals).forEach((k, v) -> observedStats.put(k, Schema.StatisticsState.maximumLengthObservedInterval, v));
+        mapIdsToMaxValue(unobservedIntervals).forEach((k, v) -> observedStats.put(k, Schema.StatisticsState.maximumLengthUnobservedInterval, v));
+        mapIdsToAvgValue(observedIntervals).forEach((k, v) -> observedStats.put(k, Schema.StatisticsState.averageLengthObservedInterval, v));
+        mapIdsToAvgValue(unobservedIntervals).forEach((k, v) -> observedStats.put(k, Schema.StatisticsState.averageLengthUnobservedInterval, v));
```
Comment (omermadmon) [inline src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:454]: Consider renaming `(k, v)` into `(beaconId, statisticValue)` (or any other meaningful names).

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java [hunk 3]
```diff
@@ -145,15 +147,24 @@ void updateSimulationStats() {
...
     /** Write final simulation statistical data to db. */
     void writeSimulationStats() {
-        for (Beacon beacon : beacons) {
-            beaconsObservedSum.putIfAbsent(String.valueOf(beacon.getId()), 0D);
-        }
-        Map<String, Double> beaconsObservedPercent = beaconsObservedSum.entrySet().stream()
-                .collect(toImmutableMap(e -> e.getKey(), e -> e.getValue() / (currentRound - 1)));
+        Map<String, List<Integer>> observedIntervals = beaconsObservedIntervals.entrySet().stream()
+                .collect(toImmutableMap(e -> String.valueOf(e.getKey().getId()),
+                e -> e.getValue().stream().filter(i -> i.observed()).map(ObservedInterval::duration).collect(toImmutableList())));
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:152]: Please split this to one more line and indent the `collect()` arguments one under the other, to make the code easier to read.

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java [hunk 4]
```diff
@@ -145,15 +147,24 @@ void updateSimulationStats() {
...
+
+        Map<String, List<Integer>> unobservedIntervals = beaconsObservedIntervals.entrySet().stream()
+                .collect(toImmutableMap(e -> String.valueOf(e.getKey().getId()),
+                e -> e.getValue().stream().filter(i -> !i.observed()).map(ObservedInterval::duration).collect(toImmutableList())));
+
+        Map<String, Double> beaconsObservedPercent = observedIntervals.entrySet().stream() // delete after deletion from StatsState
+                .collect(toImmutableMap(e -> e.getKey(), // delete after deletion from StatsState
+                e -> e.getValue().stream().mapToDouble(Double::valueOf).sum() / (currentRound - 1))); // delete after deletion from StatsState
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:160]: Consider adding a helper function for this partial sum (for better readability).

Comment (nolaviz) [reply src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:160]: (Alternatively: skip the `mapToDouble()`, summing the integer intervals as-is, and cast `(currentRound - 1)` to `(double)(currentRound - 1)` - Java will promote the integer sum to double, and the code is slightly shorter)

File: src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java [hunk 5]
```diff
@@ -404,7 +415,70 @@ protected AbstractSimulation (Builder builder) {
...
+        mapIdsToMaxValue(unobservedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.maximumLengthUnobservedInterval, value));
+        mapIdsToAvgValue(observedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.averageLengthObservedInterval, value));
+        mapIdsToAvgValue(unobservedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.averageLengthUnobservedInterval, value));
+        return observedStats.build();
+    }
+
+    private Map<String, Double> mapIdsToPercentOfSimulation(Map<String, List<Integer>> idsToIntervals) {
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java:465]: This looks like the same code as in line 158 - right? If so, (a) consider applying the same simplification I recommended there; (b) consider using this method from line 158, avoiding duplication.

File: src/main/java/com/google/research/bleth/simulator/StatisticsState.java [hunk 6]
```diff
@@ -83,6 +90,29 @@ public void writeBeaconsObservedPercentStats() {
...
+     */
+    public void writeBeaconsObservedStats() {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        Table<String, String, Double> beaconsIdToObservedStats = readBeaconsObservedStats(simulationId);
+        if (!beaconsIdToObservedStats.isEmpty()) {
+            throw new StatisticsAlreadyExistException(simulationId);
+        }
+
+        Map<String, Map<String, Double>> idToObservedStats = beaconsObservedStats.rowMap();
```
Comment (nolaviz) [inline src/main/java/com/google/research/bleth/simulator/StatisticsState.java:106]: Why do you need this change in representation?
You could just use `beaconsObservedStats.rowKeySet()` in the loop, and `beaconsObservedStats.row(beaconId)` on line 111.

**Review by nolaviz — APPROVED** [review]

**Pull request merged**